src/index.js

```diff
@@ -883,34 +929,49 @@ export default function (babel) {
       },
 
       ArrayComprehension(path) {
-        // disallow Yield and Return
-        path.get("loop.body").traverse({
-          AwaitExpression(awaitPath) {
-            throw awaitPath.buildCodeFrameError(
-              "`await` is not allowed within Comprehensions; " +
-              "instead, await the Comprehension (eg; `y <- [for x of xs: x]`)."
-            );
-          },
-          YieldExpression(yieldPath) {
-            throw yieldPath.buildCodeFrameError("`yield` is not allowed within Comprehensions.");
-          },
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
```

Comment:
Not sure this will work. Regardless of enter or exit, the visitor is being called at a point-in-time -- exactly when we call `validateComprehensionLoopBody`, so if the processing hasn't been done by then it won't matter.

A possible change here is to move the whole comprehension visitor to exit.